incubator/hnc/pkg/kubectl/set.go

```diff
@@ -16,100 +16,148 @@ limitations under the License.
 package kubectl
 
 import (
+	"errors"
 	"fmt"
 	"os"
 
 	"github.com/spf13/cobra"
+
+	api "github.com/kubernetes-sigs/multi-tenancy/incubator/hnc/api/v1alpha1"
 )
 
+//flagValues struct stores name of namespaces against type of flag passed
+type flagValues struct {
+	root                bool
+	parentString        string
+	requiredChildArray  []string
+	optionalChildString string
+}
+
 var setCmd = &cobra.Command{
 	Use:     "set",
 	Short:   "Allows setting namespace hierarchy",
 	Aliases: []string{"set"},
 	Args:    cobra.MinimumNArgs(1),
 	Run: func(cmd *cobra.Command, args []string) {
 		nnm := args[0]
-		hc := client.getHierarchy(nnm)
-		oldpnm := hc.Spec.Parent
 		flags := cmd.Flags()
-		numChanges := 0
+		parentString, _ := flags.GetString("parent")
+		requiredChildArray, _ := flags.GetStringArray("requiredChild")
+		optionalChildString, _ := flags.GetString("optionalChild")
+
+		flagValues := flagValues{
+			root:                flags.Changed("root"),
+			parentString:        parentString,
+			requiredChildArray:  requiredChildArray,
+			optionalChildString: optionalChildString,
+		}
 
-		if flags.Changed("root") && flags.Changed("parent") {
-			fmt.Println("Cannot set both --root and --parent at the same time")
+		msgs, err := setCmdFunc(client, flagValues, nnm)
+		if err != nil {
+			fmt.Println(err.Error())
 			os.Exit(1)
 		}
 
-		if flags.Changed("root") {
-			if oldpnm == "" {
-				fmt.Printf("%s is already a root namespace; unchanged", nnm)
-			} else {
-				hc.Spec.Parent = ""
-				fmt.Printf("Unsetting the parent of %s (was previously %s)\n", nnm, oldpnm)
-				numChanges++
-			}
+		for _, each := range msgs {
+			fmt.Print(each)
```

Comment:
@adrianludwin  added a comment : What's the benefit from delaying all these messages from being printed out? I don't think we should be unit testing the messages themselves, is that the only reason?

Reply:
Yes I am thinking to make our main testable function to return proper messages so that I can test them properly.

Reply:
I usually don't like testing human-readable messages because it discourages people from making changes or improvements. Maybe if kubectl-hnc gets much more complex we could do something smart (ie return a struct that's easy to compare, where one value in the struct is a pointer to a Printf format string) but I think that's overkill for now.